website/content/v1.10/advanced/kernel-module.md

```diff
@@ -0,0 +1,274 @@
+    test:             # validate software
+      - fhs-validator /rootfs
+finalize:             # copy directory structure from source to destination
+  - from: /rootfs
+    to: /
+```
+
+## Build the package and kernel
+
+After you've created a pkg.yaml file you can test building your package with the make target you generated earlier.
+Because Talos requires kernel modules to be signed with a signing key only available during the Talos kernel build process we need to build the kernel and package at the same time.
+
+We also need a container registry available to store the built assets.
+Follow the steps in [developing Talos]({{< relref "../advanced/developing-talos#prepare" >}}) to create a docker builder and run a local container registry before running this command.
+
+```bash
+make kernel my-module-pkg REGISTRY=127.0.0.1:5005 \
+  PLATFORM=linux/amd64 \
+  PUSH=true
+```
+
+If this is successful it should output two pieces of information we need to collect for the next steps.
+We need to save the kernel and package images.
+The output will look something like this:
+
+```bash
+=> => pushing manifest for 127.0.0.1:5005/user/kernel:v1.11.0-alpha.0...
+...
+=> => pushing manifest for 127.0.0.1:5005/user/my-module:v1.11.0-alpha.0...
+```
+For easier reference in this guide I will save these images as `$KERNEL_IMAGE` and `$PKG_IMAGE` variables.
+
+## Create an extension
+
+System extensions are the way to add software and files to a Talos Linux root filesystem.
+Just like packages they are built as containers and then layered with Talos to create a bootable squashfs image.
+
+The only unique thing about building a system extension with a kernel module is we need to build it against the kernel we just built in the previous step.
+If we don't do this then our kernel module won't be signed and cannot be loaded at runtime.
+
+The process is very similar to creating a package.
+Start by cloning the extensions repo:
+
+```bash
+git clone https://github.com/siderolabs/extensions
+cd extensions
+mkdir my-module
+```
+Add your extension to the `.kres.yaml` file.
+
+```yaml
+---
+kind: pkgfile.Build
+spec:
+  targets:
+    ...
+    - my-module
+    ...
+```
+Then generate a new Makefile with additional target.
+
+```bash
+make rekres
+```
+
+Now create the `manifest.yaml` file for the metadata of your extension in the my-module folder.
+```yaml
+version: v1alpha1       # version of manifest.yaml
+metadata:
+  name: my-module
+  version: 0.1
+  author: me
+  description: |
+    An extension that adds a kernel module
+  compatibility:
+    talos:
+      version: ">= v1.10.0"  # what version of Talos is supported
+```
+Create a pkg.yaml file which works similarly to our pkg.yaml file for our package, but this time starts from the base image we built in the first step.
+The local directory is mounted into the container at `/pkg` so we can copy files from that directory.
+
+```yaml
+name: my-module
+variant: scratch
+shell: /bin/sh
```

Comment:
Where does the shell come from with scratch environments? I'm trying to run through these docs myself to make sure all the examples work and getting the error
```
/bin/sh: line 1: no: No such file or directory
```
when trying to create the extension. I don't know how this part works.